File: src/main/java/com/dude/dms/backend/brain/BrainUtils.java
package com.dude.dms.backend.brain;

import java.io.*;
import java.util.Properties;

public final class BrainUtils {

    private static final DmsLogger LOGGER = DmsLogger.getLogger(BrainUtils.class);

    private BrainUtils() {
    }

    public static String getProperty(OptionKey key) {
        String userProperty = getUserProperty(key);
        return userProperty != null ? userProperty : getDefaultProperty(key);
    }

    public static void setProperty(OptionKey key, String value) {
        Properties prop = getUserProperties();
        prop.setProperty(key.key, value);
        try (OutputStream output = new FileOutputStream(new File("options.properties"))) {
            prop.store(output, null);
        } catch (IOException io) {
            LOGGER.error(io.getMessage());
        }
    }

    private static String getDefaultProperty(OptionKey key) {
        return getDefaultProperties().getProperty(key.key);
    }

    private static String getUserProperty(OptionKey key) {
        return getUserProperties().getProperty(key.key);
    }

    private static Properties getUserProperties() {
        Properties prop = new Properties();
        try (InputStream input = new FileInputStream(new File("options.properties"))) {
            prop.load(input);
        } catch (IOException ex) {
            LOGGER.error(ex.getMessage());
        }
        return prop;
    }

    private static Properties getDefaultProperties() {
        Properties prop = new Properties();
        try (InputStream input = new FileInputStream(new File("options.default.properties"))) {
            prop.load(input);
        } catch (IOException ex) {
            LOGGER.error(ex.getMessage());
        }
        return prop;
    }
}

File: src/main/java/com/dude/dms/backend/brain/BrainUtils.kt
package com.dude.dms.backend.brain

import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.util.*

object BrainUtils {

    private val LOGGER = DmsLogger.getLogger(BrainUtils::class.java)

    @JvmStatic
    fun getProperty(key: OptionKey) = getUserProperty(key) ?: getDefaultProperty(key)

    @JvmStatic
    fun setProperty(key: OptionKey, value: String?) {
        val prop = getProperties("options.properties")
        prop.setProperty(key.key, value)
        try {
            FileOutputStream(File("options.properties")).use { output -> prop.store(output, null) }
        } catch (io: IOException) {
            io.message?.let { LOGGER.error(it) }
        }
    }

    private fun getDefaultProperty(key: OptionKey) = getProperties("options.default.properties").getProperty(key.key)

    private fun getUserProperty(key: OptionKey) = getProperties("options.properties").getProperty(key.key)

    private fun getProperties(file: String): Properties {
        val prop = Properties()
        try {
            FileInputStream(File(file)).use { input -> prop.load(input) }
        } catch (ex: IOException) {
            ex.message?.let { LOGGER.error(it) }
        }
        return prop
    }
}